internal/rbd/sms_controllerserver.go

```diff
@@ -0,0 +1,220 @@
+
+	ctx := context.WithValue(stream.Context(), log.ReqID, req.GetSnapshotId())
+
+	mgr := NewManager(cs.Driver.GetInstanceID(), nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	rbdSnap, err := mgr.GetSnapshotByID(ctx, snapshotID)
+	if errors.Is(err, rbderrors.ErrImageNotFound) {
+		log.ErrorLog(ctx, "snapshot %q not found: %v", snapshotID, err)
+
+		return status.Errorf(codes.NotFound, "snapshot %q not found: %v", snapshotID, err)
+	}
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	defer rbdSnap.Destroy(ctx)
+
+	volSize := rbdSnap.GetVolSize()
+
+	startingOffset := req.GetStartingOffset()
+	if startingOffset >= volSize {
+		return status.Error(codes.OutOfRange, "starting offset is out of range")
+	}
+
+	maxResults := normalizeMaxResults(ctx, req.GetMaxResults())
+
+	sendResponse := func(changedBlocks []*csi.BlockMetadata) error {
+		return stream.Send(&csi.GetMetadataAllocatedResponse{
+			BlockMetadataType:   csi.BlockMetadataType_VARIABLE_LENGTH,
+			VolumeCapacityBytes: volSize,
+			BlockMetadata:       changedBlocks,
+		})
+	}
+
+	err = rbdSnap.StreamMetadata(ctx, sendResponse, startingOffset, maxResults, nil)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	return nil
+}
+
+func validateMetadataDeltaReq(req *csi.GetMetadataDeltaRequest) error {
+	if req.GetBaseSnapshotId() == "" {
+		return status.Error(codes.InvalidArgument, "base snapshot ID cannot be empty")
+	}
+
+	if req.GetTargetSnapshotId() == "" {
+		return status.Error(codes.InvalidArgument, "target snapshot ID cannot be empty")
+	}
+
+	if req.GetSecrets() == nil || len(req.GetSecrets()) == 0 {
+		return status.Error(codes.InvalidArgument, "secrets cannot be nil or empty")
+	}
+
+	if req.GetMaxResults() < 0 {
+		return status.Error(codes.InvalidArgument, "max results cannot be negative")
+	}
+
+	if req.GetStartingOffset() < 0 {
+		return status.Error(codes.OutOfRange, "starting offset cannot be negative")
+	}
+
+	return nil
+}
+
+// GetMetadataDelta streams the block metadata delta between two snapshots.
+func (cs *ControllerServer) GetMetadataDelta(
+	req *csi.GetMetadataDeltaRequest,
+	stream csi.SnapshotMetadata_GetMetadataDeltaServer,
+) error {
+	if err := validateMetadataDeltaReq(req); err != nil {
+		return err
+	}
+
+	baseSnapshotID := req.GetBaseSnapshotId()
+	targetSnapshotID := req.GetTargetSnapshotId()
+
+	ctx := context.WithValue(stream.Context(), log.ReqID, targetSnapshotID)
+	ctx = context.WithValue(ctx, "Base Snapshot ID", baseSnapshotID)
+
+	mgr := NewManager(cs.Driver.GetInstanceID(), nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	baseRBDSnap, err := mgr.GetSnapshotByID(ctx, baseSnapshotID)
+	if errors.Is(err, rbderrors.ErrImageNotFound) {
+		log.ErrorLog(ctx, "snapshot %q not found: %v", baseSnapshotID, err)
+
+		return status.Errorf(codes.NotFound, "snapshot %q not found: %v", baseSnapshotID, err)
+	}
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
```

Comment:
```suggestion
if err != nil {
    if errors.Is(err, rbderrors.ErrImageNotFound) {
		log.ErrorLog(ctx, "snapshot %q not found: %v", baseSnapshotID, err)

		return status.Errorf(codes.NotFound, "snapshot %q not found: %v", baseSnapshotID, err)
	}
		return status.Error(codes.Internal, err.Error())
	}

it would be nice to have one block for the error handling, take care of it in other places as well.